From 09f2477706e47d022fc3c6adcdf79b1eb6634df9 Mon Sep 17 00:00:00 2001
From: Haihua Hu <jared.hu@nxp.com>
Date: Thu, 29 Mar 2018 11:18:56 +0800
Subject: [PATCH] MMFMWK-7954 waylandsink: change to using GCond to avoid
 dropping video buffer

upstream status: imx specific
https://bugzilla.gnome.org/show_bug.cgi?id=794793

Conflicts:
	ext/wayland/gstwaylandsink.c
(cherry picked from commit a70b222d23a7b74ef92919d931d0bf0df26c3c0e)
---
 ext/wayland/gstwaylandsink.c | 11 +++++------
 ext/wayland/gstwaylandsink.h |  1 +
 2 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index 2f116bf5a..878ddc396 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -193,6 +193,7 @@ gst_wayland_sink_init (GstWaylandSink * self)
 {
   g_mutex_init (&self->display_lock);
   g_mutex_init (&self->render_lock);
+  g_cond_init (&self->redraw_wait);
 }
 
 static void
@@ -319,6 +320,7 @@ gst_wayland_sink_finalize (GObject * object)
 
   g_mutex_clear (&self->display_lock);
   g_mutex_clear (&self->render_lock);
+  g_cond_clear (&self->redraw_wait);
 
   G_OBJECT_CLASS (parent_class)->finalize (object);
 }
@@ -687,6 +689,7 @@ frame_redraw_callback (void *data, struct wl_callback *callback, uint32_t time)
 
   g_mutex_lock (&self->render_lock);
   self->redraw_pending = FALSE;
+  g_cond_signal (&self->redraw_wait);
 
   if (self->callback) {
     wl_callback_destroy (callback);
@@ -769,12 +772,8 @@ gst_wayland_sink_show_frame (GstVideoSink * vsink, GstBuffer * buffer)
     }
   }
 
-  /* drop buffers until we get a frame callback */
-  if (self->redraw_pending) {
-    GST_LOG_OBJECT (self, "buffer %" GST_PTR_FORMAT " dropped (redraw pending)",
-        buffer);
-    goto done;
-  }
+  while (self->redraw_pending)
+    g_cond_wait (&self->redraw_wait, &self->render_lock);
 
   /* make sure that the application has called set_render_rectangle() */
   if (G_UNLIKELY (gst_wl_window_get_render_rectangle (self->window)->w == 0))
diff --git a/ext/wayland/gstwaylandsink.h b/ext/wayland/gstwaylandsink.h
index 3243d8cd8..4847219f6 100644
--- a/ext/wayland/gstwaylandsink.h
+++ b/ext/wayland/gstwaylandsink.h
@@ -60,6 +60,7 @@ struct _GstWaylandSink
   gchar *display_name;
 
   gboolean redraw_pending;
+  GCond redraw_wait;
   GMutex render_lock;
   GstBuffer *last_buffer;
 
